File: 프로그래머스/lv2/64065. 튜플/튜플.cpp
#include <string>
#include <vector>
#include <map>
#include <queue>
using namespace std;

typedef pair<int, int> p;

vector<int> solution(string s) {
	vector<int> answer;
	map<int,int> tmp;
	priority_queue<p, vector<p>, less<p>> pq;

	int s_len = s.length();

	string num = "";
	for (int i = 0; i < s_len; i++){
		//숫자면 일단 이어
		if ('0' <= s[i] && s[i] <= '9') num += s[i];
		//숫자 아닌데 num에 숫자가 있으면 변환
		else {
			if (num != "") {
				int cur = stoi(num);
				tmp[cur]++;
				num = "";
			}
		}
	}

	for (auto i = tmp.begin(); i != tmp.end(); i++) {
		pq.push(make_pair(i->second, i->first));
	}

	while (!pq.empty()) {
		p cur = pq.top(); pq.pop();
		answer.push_back(cur.second);
	}

	return answer;
}

File: 프로그래머스/lv2/64065. 튜플/튜플.kt
import java.util.PriorityQueue

class Solution {
    fun solution(s: String): IntArray {
    var answer = intArrayOf()

    // 숫자 꺼내기
    val numCount = IntArray(1000001)
    var numStr = ""
    s.forEach { value ->
        // 숫자인 경우 이어
        if (value in '0'..'9'){
            numStr += value
        }
        // 끝나는 그거면 넣어
        if ((value == ',' || value == '}') && numStr.isNotEmpty()){
            numCount[numStr.toInt()]++
            numStr = ""
        }
    }
    
    // 우선순위 큐에 담기
    val pq = PriorityQueue<Pair<Int, Int>>(){ a, b -> b.second - a.second}
    repeat(1000001){
        if (numCount[it] != 0){
            pq.add(Pair(it, numCount[it]))
        }
    }
    
    while (pq.isNotEmpty()){
        answer += pq.remove().first
    }
    
    return answer
    }
}